### PR TITLE
dead_code: Properly inspect fields in struct patterns with type relative paths

### LINKED ISSUE #63151: Spurious unused field warning when matching variants under a `Self` scope
This code should compile without warnings:

```rust
enum Enum {
    Variant { field: usize }
}
impl Enum {
    fn read_field(self) -> usize {
        match self {
            // Enum::Variant { field } => field
            Self::Variant { field } => field
        }
    }
}
fn main() {
    let e = Enum::Variant { field: 42 };
    println!("{}", e.read_field());
}
```

However, it produces the following one:

```
warning: field is never used: `field`
 --> src/main.rs:2:15
  |
2 |     Variant { field: usize }
  |               ^^^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default
```

If we switch the pattern within `match` to the one that's commented out, the warning is gone.

### PR DESCRIPTION
**Pull request by jakubadamw**
Closes #63151.

### CODE REVIEW

File: src/test/ui/issues/issue-63151.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// check-pass
```
Comment (Centril) [inline src/test/ui/issues/issue-63151.rs]: Can you please move this file to `src/test/ui/type-alias-enum-variants`, give the file name a more descriptive name `issue-63151-$short_description.rs` and also leave a comment in the file?

Comment (jakubadamw) [reply src/test/ui/issues/issue-63151.rs]: @Centril, sure, done.

Comment (jakubadamw) [reply src/test/ui/issues/issue-63151.rs]: I see these rules are described in the guidelines. Sorry, I should have looked there earlier. :) I just didn't recall there being any last time I contributed here around five years ago, that's why I didn't. ;-)

Comment (Centril) [reply src/test/ui/issues/issue-63151.rs]: No problem; thanks for the change and the PR!

**Pull request merged**